### PR TITLE
Fixed #23759 -- Preserved all file extensions in Storage.get_available_name().

### PR DESCRIPTION
**Pull request by adamzap**
https://code.djangoproject.com/ticket/23759

### CODE REVIEW

File: tests/file_storage/tests.py [hunk 1]
```diff
@@ -782,6 +782,21 @@ def test_duplicate_filename(self):
             for o in objs:
                 o.delete()
 
+    def test_duplicate_filename_with_multiple_extension_suffixes(self):
```
Comment (adamzap) [inline tests/file_storage/tests.py:785]: Let me know if the test could be structured better. There's a lot of duplication here.

**Comment by nessita** [comment]
Hello! Thanks for your contribution. Please remember that when this work is ready for review, you need to adjust the ticket flags in the Trac system so this PR gets added to the "branches needing review" section of the [Django Developer Dahsboard](https://dashboard.djangoproject.com/). More information in [these docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-style).

**Comment by adamzap** [comment]
@nessita Done, thanks!

File: tests/file_storage/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/file_storage/tests.py:785]: We could use `self.SubTest()`, e.g.
```diff
diff --git a/tests/file_storage/tests.py b/tests/file_storage/tests.py
index 637de0a3c9..e9dd1389c1 100644
--- a/tests/file_storage/tests.py
+++ b/tests/file_storage/tests.py
@@ -769,18 +769,24 @@ class FileFieldStorageTests(TestCase):
 
     def test_duplicate_filename(self):
         # Multiple files with the same name get _(7 random chars) appended to them.
-        objs = [Storage() for i in range(2)]
-        for o in objs:
-            o.normal.save("multiple_files.txt", ContentFile("Same Content"))
-        try:
-            names = [o.normal.name for o in objs]
-            self.assertEqual(names[0], "tests/multiple_files.txt")
-            self.assertRegex(
-                names[1], "tests/multiple_files_%s.txt" % FILE_SUFFIX_REGEX
-            )
-        finally:
-            for o in objs:
-                o.delete()
+        tests = [
+            ("multiple_files", "txt"),
+            ("multiple_files_many_extensions", "tar.gz"),
+        ]
+        for filename, extension in tests:
+            with self.subTest(filename=filename):
+                objs = [Storage() for i in range(2)]
+                for o in objs:
+                    o.normal.save(f"{filename}.{extension}", ContentFile("Same Content"))
+                try:
+                    names = [o.normal.name for o in objs]
+                    self.assertEqual(names[0], f"tests/{filename}.{extension}")
+                    self.assertRegex(
+                        names[1], f"tests/{filename}_{FILE_SUFFIX_REGEX}.{extension}"
+                    )
+                finally:
+                    for o in objs:
+                        o.delete()
 
     def test_file_truncation(self):
         # Given the max_length is limited, when multiple files get uploaded

```

**Review by felixxm — COMMENTED** [review]
@adamzap Thanks :+1: I left a small suggestion regarding tests.

File: tests/file_storage/tests.py (hunk 1 above)
Comment (adamzap) [reply tests/file_storage/tests.py]: Thanks! I was was leaning in this direction as well.

**Comment by felixxm** [comment]
@adamzap Thanks :+1:

**Comment by tmutimer** [comment]
I can't see this change documented in the release notes, have I missed it, or should it be added?

**Pull request merged**